src/aspire/source/xform.py

```diff
@@ -0,0 +1,388 @@
+        Initialize the Filter Xform using a Filter object
+        :param filter: An object of type `aspire.utils.filters.Filter`
+        :param resolution: Resolution of images expected to pass through this Transform
+        """
+        super().__init__(resolution=resolution)
+        self.filter = filter
+
+    def _forward(self, im, indices):
+        return im.filter(self.filter)
+
+    def downsample(self, resolution):
+        self.filter.scale(self.resolution / resolution)
+        super().downsample(resolution)
+
+
+class NoiseAdder(OneWayXform):
+    """
+    A Xform that adds white noise, optionally passed through a Filter object, to all incoming images.
+    """
+    def __init__(self, resolution=np.inf, seed=0, noise_filter=None):
+        """
+        Initialize the random state of this NoiseAdder using specified values.
+        :param resolution: Resolution of images expected to pass through this Transform
+        :param seed: The random seed used to generate white noise
+        :param noise_filter: An optional aspire.utils.filters.Filter object to use to filter the generated white noise.
+            Be default, a ScalarFilter is used, emulating true white noise, but any additional filter can be used
+            to emulate isotropic noise etc.
+        """
+        super().__init__(resolution=resolution)
+        self.seed = seed
+        self.noise_filter = noise_filter or ScalarFilter(value=1)
+
+    def _forward(self, im, indices):
+        im = im.copy()
+
+        for i, idx in enumerate(indices):
+            # Note: The following random seed behavior is directly taken from MATLAB Cov3D code.
+            random_seed = self.seed + 191 * (idx + 1)
+            im_s = randn(2 * self.resolution, 2 * self.resolution, seed=random_seed)
+            im_s = Image(im_s).filter(self.noise_filter, power=0.5)[:, :, 0]
+            im[:, :, i] += im_s[:self.resolution, :self.resolution]
+
+        return im
+
+
+class IndexedXform(Xform):
+    """
+    An IndexedXform is a Xform where individual Xform objects are used at specific indices of the incoming Image object.
+
+    This extra layer of abstraction is used because individual Xform objects are typically capable of dealing with
+    a stack of images (an Image object) in an efficient manner. The IndexedXform class groups the invocation of each
+    of its unique Xform objects, so that calls to individual Xform objects within it is minimized, and equals the no.
+    of unique Xforms found.
+    """
+    def __init__(self, unique_xforms, indices=None):
+        if indices is None:
+            indices = np.arange(0, len(unique_xforms))
+        else:
+            # Ensure we're dealing with a numpy array since we'll be utilizing numpy indexing later on.
+            indices = np.array(indices)
+            assert np.min(indices) >= 0
+            assert np.max(indices) < len(unique_xforms)
+
+        resolution = min(np.inf, *[xform.resolution for xform in unique_xforms])
```

Comment:
This is no longer an issue since Xform do not maintain a track of their `resolution` anymore (which was only ever used to downsample them). If the `forward` and `adjoint` methods of an `Xform` can deal with an `Image` of a particular resolution at runtime, that's good enough.